src/main/java/com/google/sps/servlets/ChatServlet.java

```diff
@@ -0,0 +1,118 @@
+    ImmutableList<String> messagesList = Streams.stream(preparedMessageQuery.asIterable()).map(message -> 
+        (String) message.getProperty(MESSAGE_TEXT_PROPERTY)).collect(toImmutableList());
+
+    // Data will be passed in as a list of messages in a map (needed for template)
+    ImmutableMap<String, ImmutableList<String>> data = ImmutableMap.of("messages", messagesList);
+
+    // File path starts in target/portfolio-1
+    SoyFileSet sfs = SoyFileSet
+        .builder()
+        .add(new File("../../src/main/java/templates/chat.soy"))
+        .build();
+    SoyTofu tofu = sfs.compileToTofu();
+
+    final String out = tofu.newRenderer("templates.chat.chatPage").setData(data).render();
+    response.getWriter().println(out);
+  }
+
+/**
+ * Is called by a button on the chat form. doPost takes the message and its
+ * attributes (such as timestamp and user) and stores them into Datastore.
+ */
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    final String messageText = (String) request.getParameter(MESSAGE_TEXT_PROPERTY);
+    final long timestamp = System.currentTimeMillis();
+    
+    // Reads API Key and Referer from file 
+    File apiKeyFile = new File("../../keys.txt");
+    Scanner scanner = new Scanner(apiKeyFile);
+    final String API_KEY = scanner.nextLine();
+    final String REFERER = scanner.nextLine();
+    scanner.close();
+
+    final String perspectiveURL = "https://commentanalyzer.googleapis.com/v1alpha1/comments:analyze?key=" + API_KEY;
+ 
+    CloseableHttpClient httpClient = HttpClients.createDefault();
+    HttpPost httpPost = new HttpPost(perspectiveURL);
+ 
+    PerspectiveRequest messageObject = new PerspectiveRequest(messageText);
+    final String inputJson = new Gson().toJson(messageObject);
+ 
+    httpPost.setEntity(new StringEntity(inputJson));
+    httpPost.addHeader("Referer", REFERER);
+    CloseableHttpResponse httpResponse = httpClient.execute(httpPost);
+ 
+    // Perspective API responds with JSON string
+    final String result = EntityUtils.toString(httpResponse.getEntity());
+    // Parsing JSON string to Java Maps
+    Map resultMap = new Gson().fromJson(result, Map.class);
+    Map attributeScores = (Map) resultMap.get("attributeScores");
+    Map toxicity = (Map) attributeScores.get("TOXICITY");
+    Map summaryScore = (Map) toxicity.get("summaryScore");
+    final Double commentScore = (Double) summaryScore.get("value");
+
+    // If the comment is not toxic, then post it to Datastore
+    if (commentScore <= 0.85) {
+        DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+        Entity messageEntity = new Entity(MESSAGE_KIND);
+        messageEntity.setProperty(MESSAGE_TEXT_PROPERTY, messageText);
+        messageEntity.setProperty(TIMESTAMP_PROPERTY, timestamp);
+        datastore.put(messageEntity);
+    } else {
+        //TODO: display a message for the user
+    }
```

Comment:
Could you make a couple helper methods for readability purposes? Maybe one for 96-103 and one for 106-114?